### PR TITLE
refactor: replace initialize() with constructor calls

### PR DESCRIPTION
**Pull request by albinso**
Making a PR here so we can have a permanent history and code review

### CODE REVIEW

**Comment by albinso** [comment]
Two issues as far as I can tell right now.
1. When Team is first initiated the input protocol lacks some fields that are required for set_flag and set_base. This can be remedied by having a boolean parameter indicating that the methods should not be run. We should find a better long-term solution.

2. Flag pickup is broken, even with the current fix to 1. Problem looks to be that base and flag are set to None so that when distance is calculated to them an exception is thrown. See lines 234-243 in pyspades/player.py (note, not piqueserver/player.py!!!)

**Comment by albinso** [comment]
I'm pretty sure problem 2, and possibly 1, detailed above are caused by changes to the Team state that are not contained to the initialization variables we use to copy the Team. I haven't been able to make out what this would be.

**Comment by esplemea** [comment]
Will do problem 1 tomorrow morning, have an idea on how to solve it.

**Comment by Perpolo** [comment]
i think we need to copy variable other (from Team) as well, it's missing after initialize and its only set at the beginning in ServerProtocol, it seems to me it's a bug

**Comment by albinso** [comment]
My latest commit fixes both problems 1 and 2 described above. Any time teams are reset they should also be set to each others' 'other' fields and ServerProtocol.teams should be updated. I'd suggest extracting the code for this into some sort of reset_teams() method to make sure it is done properly each time.

File: piqueserver/game_modes/infiltration.py [hunk 1]
```diff
@@ -78,8 +78,8 @@ def score(self):
         intel_capture.winning = winning
         self.protocol.send_contained(intel_capture, save=True)
         if winning:
-            self.team.initialize()
-            self.team.other.initialize()
+            self.team = self.team(*self.team.get_init_values())
+            self.team.other = self.team(*self.team.other.get_init_values())
```
Comment (albinso) [inline piqueserver/game_modes/infiltration.py]: We could modify the reset_teams method to allow using it here as well. I'm thinking it's probably not gonna make the code better but its worth considering.

File: pyspades/team.py [hunk 2]
```diff
@@ -39,6 +48,8 @@ def initialize(self):
             return
         self.score = 0
         self.kills = 0
+        # Breaks test cases right now, removed for demonstration purposes.
```
Comment (albinso) [inline pyspades/team.py]: I keep removing these comments but then whenever I do anything useful I accidentally reset them :<

File: tests/pyspades/test_teams.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        team_template = Team(0, self.name, self.color, False, MockServerProtocol())
+        team_template.kills = 9
+        team_template.score = 5
+        team = Team(*team_template.get_init_values())
+        self.assertEquals(0, team.kills)
+        self.assertEquals(0, team.score)
+        self.assertEquals(team_template.name, team.name)
+        self.assertEquals(team_template.color, team.color)
+
+    def test_team_others(self):
```
Comment (albinso) [inline tests/pyspades/test_teams.py:34]: I'm thinking this test case, and the behaviour that it tests, is problematic. The other team should probably not be transferred when we make a new Team instance. We could switch this case up to make sure we don't do that.

**Comment by svebrant** [comment]
Looks very good! Just gotta fix the tests and we're set

File: tests/pyspades/test_teams.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        team_2 = Team(1, "team_2", (127, 127, 127), False, MockServerProtocol())
+        team_1.other = team_2
+        team_2.other = team_1
+        team_1 = Team(*team_1.get_init_values())
+        team_2.other = Team(*team_1.other.get_init_values())
+        self.assertEquals(team_2, team_1.other)
+
+
+# Below is dummy classes used to mock the actual ServerProtocol
+class MockServerProtocol:
```
Comment (albinso) [inline tests/pyspades/test_teams.py]: MockServerProtocol lacks map_info which is used in the Team constructor to check whether or not to initialize. Needs to be included or all tests fail.

**Review by albinso — COMMENTED** [review]
Some notes that should be looked at. Not all are changes that must be made.

File: tests/pyspades/test_teams.py (hunk 4 above)
Comment (svebrant) [reply tests/pyspades/test_teams.py]: Fixed

File: tests/pyspades/test_teams.py (hunk 3 above)
Comment (Perpolo) [reply tests/pyspades/test_teams.py:34]: lin 41 and 42 are the same as lines 81 and 82 in piqueserver/game_modes/inflitration.py I wanted to check if it works the way we want to

Comment (svebrant) [reply tests/pyspades/test_teams.py:34]: Fixed

File: piqueserver/game_modes/infiltration.py (hunk 1 above)
Comment (albinso) [inline piqueserver/game_modes/infiltration.py:82]: This will probably break the game mode. Should also make it so self.team.other.other = self.team

Comment (svebrant) [reply piqueserver/game_modes/infiltration.py:82]: Added

**Review by svebrant — APPROVED** [review]
Looks good to me.

**Pull request merged**